eth-contracts/contracts/Governance.sol

```diff
@@ -4,284 +4,350 @@ import "./service/registry/RegistryContract.sol";
-      proposal.voteMagnitudeNo,
-      proposal.numVotes
-      /** @notice - votes mapping cannot be returned by external function */
-    );
-  }
+    // ========================================= Getters =========================================
+
+    function getProposalById(uint256 _proposalId)
+    external view returns (
+        uint256 proposalId,
+        address proposer,
+        uint256 startBlockNumber,
+        address targetContract,
+        uint callValue,
+        string memory signature,
+        bytes memory callData,
+        Outcome outcome,
+        uint256 voteMagnitudeYes,
+        uint256 voteMagnitudeNo,
+        uint256 numVotes
+    )
+    {
+        require(
+            _proposalId <= lastProposalId && _proposalId > 0,
+            "Must provide valid non-zero _proposalId"
+        );
+
+        Proposal memory proposal = proposals[_proposalId];
+        return (
+            proposal.proposalId,
+            proposal.proposer,
+            proposal.startBlockNumber,
+            proposal.targetContract,
+            proposal.callValue,
+            proposal.signature,
+            proposal.callData,
+            proposal.outcome,
+            proposal.voteMagnitudeYes,
+            proposal.voteMagnitudeNo,
+            proposal.numVotes
+            /** @notice - votes mapping cannot be returned by external function */
+        );
+    }
 
-  function getVoteBySlashProposalAndVoter(uint256 _proposalId, address _voter)
-  external view returns (Vote vote)
-  {
-    require(_proposalId <= lastProposalId && _proposalId > 0, "Must provide valid non-zero _proposalId");
-    return slashProposals[_proposalId].votes[_voter];
-  }
+    function getVoteByProposalAndVoter(uint256 _proposalId, address _voter)
+    external view returns (Vote vote)
+    {
+        require(
+            _proposalId <= lastProposalId && _proposalId > 0,
+            "Must provide valid non-zero _proposalId"
+        );
+        return proposals[_proposalId].votes[_voter];
+    }
 
+    // ========================================= Private =========================================
+
+    function _executeTransaction(
+        address _targetContract,
+        uint256 _callValue,
+        string memory _signature,
+        bytes memory _callData
+    ) internal returns (bytes memory /** returnData */)
+    {
+        bytes32 txHash = keccak256(
+            abi.encode(
+                _targetContract, _callValue, _signature, _callData
+            )
+        );
+
+        bytes memory callData;
+
+        if (bytes(_signature).length == 0) {
+            callData = _callData;
+        } else {
+            callData = abi.encodePacked(bytes4(keccak256(bytes(_signature))), _callData);
+        }
+
+        // solium-disable-next-line security/no-call-value
+        (bool success, bytes memory returnData) = _targetContract.call.value(_callValue)(callData);
+        require(success, "Governance::executeTransaction:Transaction execution reverted.");
```

Comment:
what does returnData look like in a failed transaction?

Reply:
i ask bc it might be worth having it emitted in a TransactionFailed event our require is what fails it

Reply:
this is actually a great point. current code doesn't handle this case gracefully at all. i made a note and have a couple ideas. will discuss with you separately, but not a blocker for merging right now. this is same logic that compound has currently.